src/PerunDummy.hs

```diff
@@ -399,23 +458,103 @@ type ChannelSchema =
 -- open channel
 --
 
+start :: AsContractError e => OpenParams -> Contract w s e ()
+start OpenParams {..} = do
+  t <- currentTime
+  let c =
+        Channel
+          { pTimeLock = spTimeLock,
+            pSigningPKs = spSigningPKs,
+            pPaymentPKs = spPaymentPKs
+          }
+      s =
+        ChannelState
+          { channelId = spChannelId,
+            balances = spBalances,
+            version = 0,
+            final = False
+          }
+      d =
+        ChannelDatum
+          { channelParameters = c,
+            state = s,
+            time = t,
+            funding = head spBalances : tail (PlutusTx.Prelude.map (const 0) spBalances),
+            funded = False,
+            disputed = False
+          }
+      v = Ada.lovelaceValueOf $ head spBalances
+      tx = Constraints.mustPayToTheScript d v
+  ledgerTx <- submitTxConstraints (typedChannelValidator spChannelId) tx
+  void $ awaitTxConfirmed $ getCardanoTxId ledgerTx
+  logInfo @P.String $ printf "Started funding for channel %d with parameters %s and value %s" spChannelId (P.show c) (P.show v)
+
+fund :: forall w s. FundParams -> Contract w s Text ()
+fund FundParams {..} = do
+  (oref, o, d@ChannelDatum {..}) <- findChannel fpChannelId
+  logInfo @P.String $ printf "found channel utxo with datum %s" (P.show d)
+  -- TODO add more checks before funding
+  when funded $
+    throwError $ pack $ printf "can only fund channel that is not already funded"
+  let
+      newFunding = addFunding (balances state!!fpIndex) fpIndex funding
+      newDatum =
+        d
+          { funding = newFunding,
+            funded = newFunding == balances state
+          }
+      v = Ada.lovelaceValueOf $ sum newFunding
+      r = Redeemer $ PlutusTx.toBuiltinData Fund
+
+      lookups =
+        Constraints.typedValidatorLookups (typedChannelValidator fpChannelId)
+          P.<> Constraints.otherScript (channelValidator fpChannelId)
+          P.<> Constraints.unspentOutputs (Map.singleton oref o)
+      tx =
+        Constraints.mustPayToTheScript newDatum v
+          <> Constraints.mustSpendScriptOutput oref r
+  ledgerTx <- submitTxConstraintsWith lookups tx
+  void $ awaitTxConfirmed $ getCardanoTxId ledgerTx
+  logInfo @P.String $ printf "Funded %d Lovelace for party %d on channel %d" (balances state!!fpIndex) fpIndex fpChannelId
```

Comment:
List indexing using `!!` is generally frowned upon in Haskell. We might be able to use fixed sized arrays, but I would do this in a follow up. Just mentioning it here.

Reply:
Yes!